File: 2944. Minimum Number of Coins for Fruits/Solution.cpp
#include <deque>
#include <vector>
using namespace std;

class Solution {
public:
    int minimumCoins(vector<int>& prices)
    {
        vector<int> f(prices.size() + 1);
        f[0] = 0;
        deque<int> mono_queue;
        for (size_t i = 1; i <= prices.size(); ++i) {
            while (!mono_queue.empty() && mono_queue.front() < (i + 1) / 2)
                mono_queue.pop_front();
            while (!mono_queue.empty() && f[mono_queue.back() - 1] + prices[mono_queue.back() - 1] >= f[i - 1] + prices[i - 1])
                mono_queue.pop_back();
            mono_queue.emplace_back(i);
            f[i] = f[mono_queue.front() - 1] + prices[mono_queue.front() - 1];
        }
        return f.back();
    }
};

File: 2944. Minimum Number of Coins for Fruits/Solution.py
from collections import deque
from typing import List


class Solution:
    def minimumCoins(self, prices: List[int]) -> int:
        f = [0] * (len(prices) + 1)
        mono_queue = deque([])
        for i in range(1, len(prices) + 1):
            while mono_queue and mono_queue[0] < (i + 1) // 2:
                mono_queue.popleft()
            while mono_queue and f[mono_queue[-1] - 1] + prices[mono_queue[-1] - 1] >= f[i - 1] + prices[i - 1]:
                mono_queue.pop()
            mono_queue.append(i)
            f[i] = f[mono_queue[0] - 1] + prices[mono_queue[0] - 1]
        return f[-1]
